### PR TITLE
feat: add generic types to `.select()`

### LINKED ISSUE #252: Unknown return type
**Describe the bug**

> Missing variable type for 'data'. 
> 
> Try adding an explicit type, or remove implicit-dynamic from your analysis options file.

The Dart analyzer shows this error in both pieces of code below in an environment where `implicit-dynamics: false` is set. It is impossible to even run the app because of this error. I cannot use supabase 1.0.0 or above until it is fixed.

Code from README of the supabase package:

```dart
main() {
  final supabase = SupabaseClient('supabaseUrl', 'supabaseKey');

  // Select from table `countries` ordering by `name`
  final data = await supabase
      .from('countries')
      .select()
      .order('name', ascending: true);
}
```

Code from README of the postgrest package:

```dart
final url = 'https://example.com/postgrest/endpoint';
final client = PostgrestClient(url);
final response = await client.from('users').select();
```

**To Reproduce**

1. Enable the strong mode by adding some settings to `analysis_options.yaml`. (See the bottom of this report.)
2. Paste the code above in the editor of an IDE.
3. See the error.

**Expected behavior**

There is no error mentioned above.

**Screenshots**

supabase_bug

**Smartphone (please complete the following information):**
 - Version [e.g. 1.0.0-dev.4]
     - What do "Smartphone" and "1.0.0-dev.4" mean here? If it's about the package version, I found this bug in supabase 1.0.1.

**Additional context**

- Packages should be developed with stricter rules about types, otherwise its users will likely get into trouble.
    - e.g.
       ```yaml
       analyzer:
         strong-mode:
           implicit-casts: false
           implicit-dynamic: false
       ```
- I'm currently experiencing this error in postgrest, but I think it would be better to see if other supabase packages have the same issue and improve them if any.

### PR DESCRIPTION
**Pull request by Vinzent03**
The idea is to pass a type to ~~`from`~~ `select`, which sets the response type. I had to add another generic to `PostgrestBuilder` to properly type the converter. The type from `from` is used as parameter type for `converter`. This doesn't infer the type depending on `single` or `count`, but makes it easier to type if you know what you get. `insert`/`update`/`upsert`/`delete` now return `void` and need a `select` to change the type.

close https://github.com/supabase-community/supabase-flutter/issues/252

### CODE REVIEW

**Review by dshukertjr — COMMENTED** [review]
This looks amazing! It seems like this PR is still in Draft mode, but is it completed? 

Also, maybe you plan to do this already, but if you could write how the public API is going to change (example code on how you can pass types to `from`), that would be amazing!

**Comment by Vinzent03** [comment]
I converted it to draft to adapt the tests to the new system.

**Comment by dshukertjr** [comment]
Looking forward to this one! I really like the new API!

```dart
final data = await client
    .from<List<Map<String, dynamic>>>('countries')
    .select()
    .withConverter<List<Country>>(
        (data) => data.map(Country.fromJson).toList());
```

**Comment by Vinzent03** [comment]
It gets even better by adding typedefs for common return types.

**Comment by Vinzent03** [comment]
Okay turns out this is much more complex than I thought, because you can't cast a `List<dynamic>` to `List<String>`, but you have to use `List.from()`. I forgot that first. In addition, `T` is currently often used as type for `PostgrestResponse.data`, but `T` is now the overall return type, including `PostgrestResponse`. I'm still trying to figure things out, but this might not be as easy as first thought.

**Comment by Vinzent03** [comment]
@dshukertjr Should be working now. You can only use the specific typedefs, because we have to manually cast the data to the type. I think we should explain under which circumstances they can expect which type. Where should we add this? As well, under `from`?

**Comment by Vinzent03** [comment]
The case of insert/update/delete without `select` is still not possible, because `Null` is always inferred to `dynamic>`. https://github.com/dart-lang/language/issues/436

# Another proposal
We could make it void though, but I don't know if this is a breaking change.

We could make `from` to return `void` by default and only allow the generic on `select`, because without `select` it's always `void`/`null`, right? Returning always `null` is a bit useless, but I don't know if we want to make this `void`.

```dart
final data = await postgrest.from('users').insert({'username': 'bar'});
//data is void
```


```dart
final data = await postgrest.from('users').insert({'username': 'bar'}).select<PostgrestList>();
//data is of type PostgrestList
```

**Comment by dshukertjr** [comment]
>We could make from to return void by default and only allow the generic on select, because without select it's always void/null, right? Returning always null is a bit useless, but I don't know if we want to make this void.

I love this. Let's do it. I don't think it's a breaking change.

File: README.md
Comment (dshukertjr) [inline README.md]:
```suggestion
final response = await client
    .from<PostgrestList>('users')
    .select()
    .withConverter((data) => data.map(User.fromJson).toList());
```

**Review by dshukertjr — COMMENTED** [review]
>You can only use the specific typedefs, because we have to manually cast the data to the type. I think we should explain under which circumstances they can expect which type. Where should we add this? As well, under from?

<strike>Frankly, I'm not sure if I'm a huge fan of introducing several new typedefs in a public API. I think it could cause more confusions than solving it. Could we just do something like this:</strike>

```dart
final data = await supabase.from<List<Map<String, dynamic>>>('table').select();
```

## Edit

I had a misunderstanding on how typedef works, and noticed that above code is valid with the current codebase!

File: lib/src/postgrest.dart
Comment (dshukertjr) [inline lib/src/postgrest.dart]: We can probably add [this](https://github.com/supabase-community/supabase-flutter/issues/252#issuecomment-1288382397) suggestion here!

```suggestion
  PostgrestQueryBuilder<T> from<T extends Object?>(String table) {
```

Comment (dshukertjr) [reply lib/src/postgrest.dart]: Also, how about adding assertion here with a helpful error message here so that users can only pass the 5 allowed types here.

```dart
assert(T == PostgrestList || T == PostgrestMap || T == Nullable<PostgrestMap> || T == PostgrestListResponse || T == PostgrestMapResponse);
```

I think we can make `_Nullable` public and hide it from the export.

Comment (Vinzent03) [reply lib/src/postgrest.dart]: Adding an assertion is a good idea!

**Comment by Vinzent03** [comment]
Great, will push the new proposal later today.

File: lib/src/postgrest_query_builder.dart
Comment (dshukertjr) [inline lib/src/postgrest_query_builder.dart:54]: I was going to suggest to extend `Object`, but this will break in a case user is doing this huh? I wonder if there is a better way of solving this issue.

```dart
final data = await supabase.from('table').select();
final firstRow = data.first; // The getter 'first' isn't defined for the type 'Object'.
```

```suggestion
  PostgrestFilterBuilder<R> select<R extends Object>([
```

File: lib/src/postgrest_query_builder.dart
Comment (dshukertjr) [inline lib/src/postgrest_query_builder.dart]: I think it's better to avoid introducing new symbols to the users as much as possible even though it ends up being kind of long. I think it will be easier to understand. 

```suggestion
  /// Allowed types for [R] are:
  /// - [List<Map<String, dynamic>>]
  /// - [Map<String, dynamic>]
  /// - [Map<String, dynamic>?]
  /// - [PostgrestResponse<List<Map<String, dynamic>>>]
  /// - [PostgrestResponse<Map<String, dynamic>>]
  /// - [PostgrestResponse<Map<String, dynamic>?>]
```

**Review by dshukertjr — COMMENTED** [review]
So it looks like the final api we have is like this?

```dart
final data = await supabase.from('table').select<List<Map<String, dynamic>>>();
```

At first, I thought this might be a crazy idea, but I wonder if [this](https://github.com/supabase-community/supabase-flutter/issues/252#issuecomment-1287821313) would be a better solution for this issue that we are trying to solve. 

We could define two more classes `PostgrestListResponse` and `PostgrestMapResponse` that implements List and Map respectively. 

```dart
class PostgrestListResponse extends PostgrestResponse implements List<Map<String, dynamic>> {
  ...
}

class PostgrestMapResponse extends PostgrestResponse implements Map<String, dynamic> {
  ...
}
```

With this, we could 

```dart
final data = await supabase.from('table').select(); // data here has `PostgrestListResponse`

final firstRow = data.first; // This is valid since `PostgrestListResponse` implements List<Map<String, dynamic>>
```

**Comment by Vinzent03** [comment]
@dshukertjr You are right, this may work. I would still keep insert/... as `PostgrestResponse<void>` then. The default generic from `select` would be `PostgrestListResponse`. `single` and `maybeSingle` would change the generic to `PostgrestMapResponse`. But the methods define the generic, and the user won't be able to specify any generic themselves, right? Will have to update all the tests again then... 🤣

**Comment by Vinzent03** [comment]
How should we handle `maybeSingle`. It returns a `PostgrestResponse`, but with `data` being null. I can't return `null` from e.g. `containsKey()`, because that's not a correct override. We could throw an exception, but that's not helpful.

**Comment by dshukertjr** [comment]
@Vinzent03 It might be faster to discuss what to do with this PR. I have sent you a message on Discord. If you have time, let's discuss over a voice chat it there!

**Review by dshukertjr — APPROVED** [review]
After further thinking about it, I don't have any good idea on how we might be able to handle `maybeSingle` without a breaking change. I am happy to merge this one if you are! Thanks for the amazing work!

File: lib/src/postgrest_query_builder.dart
Comment (Vinzent03) [reply lib/src/postgrest_query_builder.dart]: What was our conclusion here yesterday? I would use your suggested change, so that the actual type is not hidden.

**Comment by dshukertjr** [comment]
@Vinzent03 Wanted to get one last approval to double check. I'm good with merging this one, but would it be okay on your end as well?

**Comment by Vinzent03** [comment]
@dshukertjr I've written a comment here. https://github.com/supabase-community/postgrest-dart/pull/94#discussion_r1006620314 I think we should change the doc more to demonstrate the "danger" of using the wrong type and that it's optional.

File: lib/src/postgrest_query_builder.dart
Comment (dshukertjr) [reply lib/src/postgrest_query_builder.dart]: @Vinzent03 
Sorry, I missed this one. Let's go with the suggested change since even though it's a bit long!

File: lib/src/postgrest_transform_builder.dart
Comment (dshukertjr) [inline lib/src/postgrest_transform_builder.dart]: I suppose we should update this one as well!
```suggestion
  /// - [List<Map<String, dynamic>>]
  /// - [Map<String, dynamic>]
  /// - [Map<String, dynamic>?]
  /// - [PostgrestResponse<List<Map<String, dynamic>>>]
  /// - [PostgrestResponse<Map<String, dynamic>>]
  /// - [PostgrestResponse<Map<String, dynamic>?>]
```

**Comment by dshukertjr** [comment]
>I think we should change the doc more to demonstrate the "danger" of using the wrong type and that it's optional.

We could probably provide some example code for each return type with some description. Something like this maybe?

```dart
  /// - [List<Map<String, dynamic>>] is returned from query without `.single()` or `maybeSingle()` and does not contain `count` option.
  ///   ```dart
  ///   final data = await supabase.from('table').select();
  ///   ```
  /// - [Map<String, dynamic>] is returned from query with `.single()` and does not contain `count` option.
  ///   ```dart
  ///   final data = await supabase.from('table').select().single();
  ///   ```
  ... etc
```

**Comment by Vinzent03** [comment]
@dshukertjr I think the types are now properly documented on both `select` methods. What do you think?

**Review by dshukertjr — APPROVED** [review]
Love the docs! Let's merge this one in! Thanks for the amazing update!

**Pull request merged**
